lib/elixir/lib/system.ex

```diff
@@ -637,6 +637,37 @@ defmodule System do
     end
   end
 
+  @doc """
+  Similar to `System.cmd/3` but raises an error on non-zero exit status.
+
+  ## Examples
+
+      iex> System.cmd! "echo", ["hello"]
+      "hello\n"
+
+  """
+  @spec cmd!(binary, [binary], keyword) :: Collectable.t()
+  def cmd!(command, args, opts \\ []) when is_binary(command) and is_list(args) do
+    case cmd(command, args, opts) do
+      {result, 0} ->
+        result
+
+      {"", status} ->
+        raise "command exited with status #{inspect(status)}"
+
+      {result, status} when is_binary(result) ->
+        result =
+          result
+          |> String.split("\n")
+          |> Enum.map_join("\n", &("        " <> &1))
+
+        raise "command exited with status #{inspect(status)}\n\n    Result:\n#{result}"
```

Comment:
Should we include the output in the exception message? This can be very long for some commands.

Reply:
I think not having the output will make this cmd!/3 much less useful but I do agree it would be simpler and in some cases could be more desirable. In this particular case, the idea was to make the exception still reasonable to print by adding new lines and whitespaces, e.g.:

```
iex(1)> System.cmd!("elixir", ["-e", "Supervisor.count_children(1)"], stderr_to_stdout: true)
** (RuntimeError) command exited with status 1

    Result:
        ** (FunctionClauseError) no function clause matching in GenServer.whereis/1

            The following arguments were given to GenServer.whereis/1:

                # 1
                1

            Attempted function clauses (showing 6 out of 6):

                def whereis(+pid+) when -is_pid(pid)-
                def whereis(+name+) when -is_atom(name)-
                def whereis(-{:global, name}-)
                def whereis(-{:via, mod, name}-)
                def whereis(-{name, local}-) when -is_atom(name)- and -local == node()-
                def whereis(-{name, node} = server-) when -is_atom(name)- and -is_atom(node)-

            (elixir) lib/gen_server.ex:1001: GenServer.whereis/1
            (elixir) lib/gen_server.ex:819: GenServer.call/3
            (elixir) lib/supervisor.ex:981: Supervisor.count_children/1
            (stdlib) erl_eval.erl:670: :erl_eval.do_apply/6
            (elixir) lib/code.ex:192: Code.eval_string/3

    (elixir) lib/system.ex:664: System.cmd!/3
```